### PR TITLE
fix #1953 Add Exceptions.isTraceback

### LINKED ISSUE #1953: Add an official mean of identifying traceback exceptions
# Motivation
There is no good way to detect if a `Throwable` is a traceback, which could be useful in cases where eg. such a traceback is added to a `CompositeException`. 

## Desired solution
Add an official API to check if a `Throwable` is a traceback.

## Considered alternatives
The `OnAssemblyException` being package-private means that `instanceof` cannot be used. Relying on `getClass().getCanonicalName()` from within user code is an alternative, but less than ideal compared to doing that from core codebase.

Another alternative would be to have a parameter to `Exceptions.unwrap` to chose wether or not to filter out tracebacks, or having it always filter out tracebacks.

## Additional context
`checkpoint` operator adds the traceback as a suppressed exception. `CompositeException` uses the same mechanism to gather its composites, so the traceback appears in the list when unwrapping the composite.

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/Exceptions.java [hunk 1]
```diff
@@ -298,6 +298,20 @@ public static boolean isMultiple(@Nullable Throwable t) {
...
+	 * @param t the {@link Throwable} to check, {@literal null} always yields {@literal false}
+	 * @return true if the Throwable is a backtrace, false otherwise
+	 */
+	public static boolean isBacktrace(@Nullable Throwable t) {
+		if (t == null) {
+			return false;
+		}
+		//FIXME maybe add an interface here for detection purposes
+		return "reactor.core.publisher.FluxOnAssembly.OnAssemblyException".equals(t.getClass().getCanonicalName());
+	}
```
Comment (rstoyanchev) [inline reactor-core/src/main/java/reactor/core/Exceptions.java]: I wonder about aligning terminology. The Javadoc for `checkpoint` talks about "assembly tracing" and "assembly traceback" vs the `backtrace` used here. Perhaps `isCheckpointError` or `isCheckpontTraceback"?

File: reactor-core/src/test/java/reactor/core/ExceptionsTest.java [hunk 2]
```diff
@@ -320,4 +321,23 @@ public void failWithRejectedReactorREENoOp() {
...
+		Mono<Throwable> mono = Mono.zipDelayError(errorMono1, errorMono2)
+		                           .checkpoint("checkpoint")
+		                           .<Throwable>map(tuple -> new IllegalStateException("should have failed: " + tuple))
+		                           .onErrorResume(Mono::just);
+		Throwable test = mono.block();
+
+		List<Throwable> exceptions = Exceptions.unwrapMultiple(test);
+
+		assertThat(exceptions).as("unfiltered composite has traceback").hasSize(3);
+		assertThat(exceptions.stream().filter(e -> !Exceptions.isTraceback(e)))
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/ExceptionsTest.java]: FYI there is also `filteredOn`:
```java
assertThat(exceptions).filteredOn(e -> !Exceptions.isTraceback(e)))
```

File: reactor-core/src/test/java/reactor/core/ExceptionsTest.java [hunk 3]
```diff
@@ -320,4 +321,23 @@ public void failWithRejectedReactorREENoOp() {
...
+
+	@Test
+	public void unwrapMultipleWithTracebackAndFilterItOut() {
+		Mono<String> errorMono1 = Mono.error(new IllegalStateException("expected1"));
+		Mono<String> errorMono2 = Mono.error(new IllegalStateException("expected2"));
+		Mono<Throwable> mono = Mono.zipDelayError(errorMono1, errorMono2)
+		                           .checkpoint("checkpoint")
+		                           .<Throwable>map(tuple -> new IllegalStateException("should have failed: " + tuple))
+		                           .onErrorResume(Mono::just);
+		Throwable test = mono.block();
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/ExceptionsTest.java]: nit: could you a better name ;)

File: reactor-core/src/test/java/reactor/core/ExceptionsTest.java [hunk 4]
```diff
@@ -320,4 +321,23 @@ public void failWithRejectedReactorREENoOp() {
...
+
+		List<Throwable> exceptions = Exceptions.unwrapMultiple(test);
+
+		assertThat(exceptions).as("unfiltered composite has traceback").hasSize(3);
+		assertThat(exceptions.stream().filter(e -> !Exceptions.isTraceback(e)))
+				.as("filter out tracebacks")
+				.hasSize(2)
+				.allMatch(e -> e instanceof IllegalStateException, "is IllegalStateException");
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/ExceptionsTest.java]:
```suggestion
				.hasOnlyElementsOfType(IllegalStateException.class);
```

**Comment by simonbasle** [comment]
@rstoyanchev does this look good? (cf the added `unwrapMultipleExcluding` helper method)

**Comment by simonbasle** [comment]
@smaldini we used the `traceback` lingo everywhere in the documentation, we might as well stay consistent.

**Comment by rstoyanchev** [comment]
I think `unwrapMultipleExcluding` would be better if it was flipped so the Predicate is to include rather than to exclude, and then it could just be an overloaded `unwrapMultiple`. That said I also wonder if the Predicate is obscuring slightly what is possibly the only reason for this. Maybe `unwrapMultipleExcludingTraceback(Throwable)` could be all that's needed?

**Comment by simonbasle** [comment]
@rstoyanchev good suggestion, I've updated the PR accordingly

**Comment by simonbasle** [comment]
(I've rebased and squashed the commits that pertain to isTraceback)

**Pull request merged**